### PR TITLE
Feature/service crypto setup

### LINKED ISSUE #23: Add credit crypto protocol to credit service

### LINKED ISSUE #24: Add issue crypto protocol to issue service

### PR DESCRIPTION
**Pull request by this-kramer**
## Reason for this PR:

- We need to figure out a setup for deploying the services
- We need to integrate crypto into our services
  - Closes #23, Closes #24

## Changes in this PR:

- Add info-service that generates pp an provider keys and has endpoints for querying them
- Adapt docker-compose to work with nginx reverse proxy
- Add crypto and communication with info service to issue service and credit service
- Add new fullworkflow integration test to client library that runs issue-join and then credit-earn

## Checklist:

Select whatever applies:

- [x] Test written
- [x] Code Properly Formatted
- [x] Documented
- [x] Changelog updated

### CODE REVIEW

File: services/client/src/main/java/org/cryptimeleon/incentive/client/dto/JoinRequestDto.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package org.cryptimeleon.incentive.client.dto;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class JoinRequestDto {
```
Comment (pschuermann97) [inline services/client/src/main/java/org/cryptimeleon/incentive/client/dto/JoinRequestDto.java:10]: what does dto stand for?

File: services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/AliveTest.java [hunk 2]
```diff
@@ -2,45 +2,56 @@
...
+
+import java.time.Duration;
 
 import static org.assertj.core.api.Assertions.assertThat;
 
 public class AliveTest extends IncentiveSystemIntegrationTest {
 
-    /*
+    /**
      * Integration test that queries the hello world endpoint
```
Comment (pschuermann97) [inline services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/AliveTest.java:16]: Comment outdated or am I missing something?

File: services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/FullWorkflowTest.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    private Logger logger = LoggerFactory.getLogger(FullWorkflowTest.class);
+
+    @Test
+    void runFullWorkflow() {
+        var infoClient = new InfoClient(infoUrl);
+        var issueClient = new IssueClient(issueUrl);
+        var creditClient = new CreditClient(creditUrl);
+        var basketClient = new BasketClient(basketUrl);
+
+        logger.info("Retrieve data from issue service");
```
Comment (pschuermann97) [inline services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/FullWorkflowTest.java:35]: info service?

File: services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/FullWorkflowTest.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        var serializedPublicParameters = infoClient.querySerializedPublicParameters().block(Duration.ofSeconds(1));
+        var serializedProviderPublicKey = infoClient.querySerializedProviderPublicKey().block(Duration.ofSeconds(1));
+
+        logger.info("Deserialize data and setup incentive system");
+        var jsonConverter = new JSONConverter();
+        var publicParameters = new IncentivePublicParameters(jsonConverter.deserialize(serializedPublicParameters));
+        var providerPublicKey = new ProviderPublicKey(jsonConverter.deserialize(serializedProviderPublicKey), publicParameters.getSpsEq(), publicParameters.getBg().getG1());
```
Comment (pschuermann97) [inline services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/FullWorkflowTest.java:42]: line breaks for readability?

File: services/info/src/main/java/org/cryptimeleon/incentive/services/info/InfoService.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    private IncentivePublicParameters pp;
+    private ProviderKeyPair providerKeyPair;
+
+    @Value("${provider.shared-secret}")
+    private String sharedSecret;
+
+    @PostConstruct
+    public void init() {
+        // Make sure shared secret is set
+        if (sharedSecret.equals("")) {
```
Comment (pschuermann97) [inline services/info/src/main/java/org/cryptimeleon/incentive/services/info/InfoService.java:37]: can it be null?

File: services/info/src/test/java/org/cryptimeleon/incentive/services/info/HttpApiTest.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.context.SpringBootTest.WebEnvironment;
+import org.springframework.test.web.reactive.server.WebTestClient;
+
+
+@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
+public class HttpApiTest {
```
Comment (pschuermann97) [inline services/info/src/test/java/org/cryptimeleon/incentive/services/info/HttpApiTest.java:16]: purpose of this class?
maybe a comment would be nice

**Review by pschuermann97 — APPROVED** [review]
Some minor things confusing me, but LGTM overall.

File: services/info/src/main/java/org/cryptimeleon/incentive/services/info/InfoService.java (hunk 5 above)
Comment (this-kramer) [reply services/info/src/main/java/org/cryptimeleon/incentive/services/info/InfoService.java:37]: No, `@Value` ensures it is not null

File: services/client/src/main/java/org/cryptimeleon/incentive/client/dto/JoinRequestDto.java (hunk 1 above)
Comment (this-kramer) [reply services/client/src/main/java/org/cryptimeleon/incentive/client/dto/JoinRequestDto.java:10]: Data Transfer Object

File: services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/AliveTest.java (hunk 2 above)
Comment (this-kramer) [reply services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/AliveTest.java:16]: No, the `/` endpoint returns with "Hello from ... service!", hence I called it hello world endpoint :)
The purpose is to have a simple test that checks whether the service is reachable under the correct url.

File: services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/FullWorkflowTest.java (hunk 4 above)
Comment (this-kramer) [reply services/client/src/test/java/org/cryptimeleon/incentive/client/integrationtest/FullWorkflowTest.java]: Fixed :+1:

File: services/info/src/test/java/org/cryptimeleon/incentive/services/info/HttpApiTest.java (hunk 6 above)
Comment (this-kramer) [reply services/info/src/test/java/org/cryptimeleon/incentive/services/info/HttpApiTest.java]: I refactored the info service tests :)

**Pull request merged**